### PR TITLE
adds virtual space at the bottom of the editor

### PR DESCRIPTION
**Pull request by draganescu**
## Description
Moves #14852 forward by adding a simple clickable space at the bottom of the editor. For me it feels amazing already! :)

## How has this been tested?
It is a very small change, I tested locally only.

## Types of changes
CSS update to the min height of the click redirect element.

## Checklist:
- [x] My code is tested.

### CODE REVIEW

**Comment by jasmussen** [comment]
Thank you for doing this! 

I'm really happy to see that this fix really WAS as simple as adding a min-height. That's great.

However in the current implementation, 60vh is probably not ideal. On a a tall browser, there's A LOT of space at the bottom. But similary because the vh unit is a percentage of the viewport height, the scrollspace varies based on how tall your viewport is, which feels inconsistent. 

space

One of the tricky things here is that we use some flex rules to make sure that on a blank page, the click redirector takes up the full vertical height until such a time as content fills up. Then the 50px base rule kicks in. Which incidentally also means that height actually _acts_ as a min-height, so regardless of what we do we need to change that instead of add a min-height. I tried 240px and that feels pretty good to me:

```
.edit-post-visual-editor .block-editor-writing-flow__click-redirect {
	// Collapse to minimum height of 240px, to fully occupy editor bottom pad.
	height: 240px;
	width: 100%;
	// Offset for: Visual editor padding, block (default appender) margin.
	margin: #{ -1 * $block-spacing } auto -50px;
}
```

With lots of contents, it looks like this:

Screenshot 2019-04-25 at 10 28 30

And inversely on a small viewport it doesn't force a scrollbar until you get to this point:

Screenshot 2019-04-25 at 10 28 54

That feels like a good balance to me. Extra scrollspace, but not enough to go crazy.

But given this appender is subject to some advanced flexing rules, I would appreciate @aduth thoughts on this, not just because he's the flexmaster, but I believe he worked on this.

**Comment by gziolo** [comment]
@draganescu - it's easier to triage PRs when there are labels assigned.

**Comment by draganescu** [comment]
Indeed @jasmussen the problem with min height set in the `vh` unit is that it forces the scrollbar to be present even when it is not the case and that it adds a giant scrolling space on larger resolutions.

However my thinking was more about the UX issue that we try to solve:

Untitled - reason

Because of that, I figured `vh` would be precisely consistent between resolutions and their varying vertical space as illustrated below:

Untitled - VH

Using `vh` the segment of the screen where the eyes would rest when editing the "end" of the content is the same distance from the top of the screen on all resolutions.

Using `px`, the higher the resolution the lower we stare at the screen:

Untitled - PX

However given enough "px" the difference might not be that bad and perhaps the useless scroll is worse than fringe cases of people with 5k displays at native resolution staring at the bottom of the editor :) With that in mind I would up the `240px` value a bit to `300px` so that the end of the content can be just above the middle of the screen most of the time. 

However I haven't figured out why, but it appears that not any px value has the same effect, e.g. between `300px` and `350px` there is no difference in the scroll height, maybe @aduth can tell us why.

**Comment by jasmussen** [comment]
That was a very lovely argument. I'm not completely convinced of it, I feel like it is important that there is consistency in how much additional space is present at the bottom, moreso than the eyeline arguments, otherwise I worry a user might ask: why is there less padding when my browser is short, than when it is long?

As to how much buffer there should be, I'm personally open too. But there's a balance to be struck here too, because due to how the min-height works, we have to also test on a completely blank page (Add new): if the min-height is too tall, it will cause a scrollbar even there, which feels off.

Adding a label to get some additional eyes on it :) Thanks again for looking at this.

**Comment by aduth** [comment]
> But given this appender is subject to some advanced flexing rules, I would appreciate @aduth thoughts on this, not just because he's the flexmaster, but I believe he worked on this.

I'd need to test further, but I seem to recall there was some specific need to align the `padding` value from `.edit-post-visual-editor` with the `height` value of `.block-editor-writing-flow__click-redirect`. As far as it impacts the changes here too, I'd wonder if a change ought to be assigned as replacing the existing `height` style anyways, not assigning a new, separate `min-height` style.

**Comment by jasmussen** [comment]
> As far as it impacts the changes here too, I'd wonder if a change ought to be assigned as replacing the existing height style anyways, not assigning a new, separate min-height style.

Yep, I reached the same conclusion in my suggestion here: 
https://github.com/WordPress/gutenberg/pull/15165#issuecomment-486573474

**Comment by draganescu** [comment]
@jasmussen I have replaced min-height of 60vh with a height of 240px.

**Comment by jasmussen** [comment]
Thanks for doing that! I think it's a solid improvement and it doesn't preclude further explorations on this. For example I think we should still consider adding some typewriter behavior in the future.

As it is now, from my point of view this is solid and good to go. 👍👍

Because of the flex rules mentioned, I'd love a sanity check and test by someone else than me, for example @kjellr, Andrew as mentioned, or @gziolo, just to make sure we don't regress anything. We should be sure to test with and without metaboxes, with short and long posts, and ideally in a few browsers. 

Thanks again!

File: packages/edit-post/src/components/visual-editor/style.scss [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 .edit-post-visual-editor .block-editor-writing-flow__click-redirect {
 	// Collapse to minimum height of 50px, to fully occupy editor bottom pad.
-	height: 50px;
+	height: 240px;
```
Comment (aduth) [inline packages/edit-post/src/components/visual-editor/style.scss]: Mentioned in https://github.com/WordPress/gutenberg/pull/15165#issuecomment-487095968 , I sense the value here was intentionally aligned to the `50px` in:

https://github.com/WordPress/gutenberg/blob/b9e2e893f6a0ff30545869466a6f3c6892ad502c/packages/edit-post/src/components/visual-editor/style.scss#L3

https://github.com/WordPress/gutenberg/blob/b9e2e893f6a0ff30545869466a6f3c6892ad502c/packages/edit-post/src/components/visual-editor/style.scss#L15

Where, if we were to change the value, the others ought similarly be updated.

I don't recall if or what specific difference it might make, or if even the visual editor needs _any_ bottom padding when the click redirect is present.

Comment (jasmussen) [reply packages/edit-post/src/components/visual-editor/style.scss]: I actually tried changing those 50px to -240px to match, and the result got weird. I don't fully understand what those 50px margin do, and I notably don't understand why it got weird. I suppose that's worth examining, otherwise I can record a GIF of what happens tomorrow. Heading out for the night now.

Comment (draganescu) [reply packages/edit-post/src/components/visual-editor/style.scss]: updating the margin to correlate with the height works all right (I have no idea what the effect is). However the padding at line 3 is the padding that pushes content below the top bar. That should be left as it is @aduth

File: packages/edit-post/src/components/visual-editor/style.scss [hunk 2]
```diff
@@ -9,10 +9,10 @@
...
 	width: 100%;
 	// Offset for: Visual editor padding, block (default appender) margin.
-	margin: #{ -1 * $block-spacing } auto -50px;
+	margin: #{ -1 * $block-spacing } auto -240px;
```
Comment (talldan) [inline packages/edit-post/src/components/visual-editor/style.scss]: The `-50px` looks like it's present because the parent of the 'click-redirect' element (`.edit-post-visual-editor`) has a 50px padding at the top and bottom:
https://github.com/WordPress/gutenberg/blob/1385adf98197a4f12b1490c87a3487ffad2c88a3/packages/edit-post/src/components/visual-editor/style.scss#L3

The `-50px` margin that was previously specified essentially counteracts that and pulls the 'click-redirect' element down to the bottom of the screen. Not sure why it continues to work when it's `-240px`, but `-50px` is technically the correct value.

TLDR, sorry to ask you to change this back again, but it looks like it should stay at 50px.

**Comment by jasmussen** [comment]
I was kind of worried that it wouldn't be as simple as this :( 

I do agree with @talldan about his comment in https://github.com/WordPress/gutenberg/pull/15165#discussion_r281905826, so the revert of that change was good. I suggested as much in this comment too: https://github.com/WordPress/gutenberg/pull/15165#discussion_r280196005

However in practice, there's something about using the _height_ instead of the _min-height_ that isn't working as it should. Watch:

changes

I'm not sure how to proceed, and would really appreciate some more looks by @aduth. Otherwise, I can take a stab at this later this week. Given that changing the 50px height to 240px doesn't do anything when you have lots of contents, I'm now actually confused why it was there in the first place.

**Comment by talldan** [comment]
@jasmussen it's a little bit hard to see what's happening in the gif, maybe I need any eye test 😄 

Is it the double scrollbar that's the issue?

**Comment by jasmussen** [comment]
No, the issue is that there is _no_ additional scrollspace at the bottom of the post, like this PR intended to add. That scrollspace is only there if we use a min-height, instead of a height.

**Comment by draganescu** [comment]
@jasmussen @talldan I reverted to using `min-height` and the bug in the gifs above seems fixed (at least for me). Can we proceed like this, merge and discuss how things work in an issue instead?

**Comment by jasmussen** [comment]
I haven't forgotten this branch, but I do apologize for not returning soon enough. 

We need to fix this at the source, though — I'm afraid the min-height fix won't work as it creates dual scrollbars:

Screenshot 2019-05-28 at 11 26 45

I believe that this PR is 90% there, and as soon as I get a breather, I hope to take a stab at a variant that incorporates your learnings here.

**Comment by jasmussen** [comment]
Actually, scratch my previous comment. I do apologize, that extra scrollbar was not the fault of this PR!

In exploring a different solution, I found that the following will work for us:

```
.edit-post-visual-editor .block-editor-writing-flow__click-redirect {
	// Collapse to minimum height of 50px, to fully occupy editor bottom pad.
	height: 50px;
	width: 100%;

	// Provide a bottom margin for some minimum scroll space.
	min-height: 160px;

	// Offset for: Visual editor padding, block (default appender) margin.
	margin: #{ -1 * $block-spacing } auto -50px;
}
```

This is essentially the best of both worlds. We don't mess with the 50px which are necessary for the delicate flex rules. But it also adds a min-height. But it's more conservative:

Screenshot 2019-05-28 at 11 34 52

It is less likely to cause issues on small screens. Can you try that code above? If you dig it, then I'd love a sanity check from @kjellr  and then ship it afterwards!

**Comment by draganescu** [comment]
Yep the solution is quite good and works pretty nice, it is also up in this PR :)  @kjellr let us know of any observations!

**Comment by jasmussen** [comment]
Thank you for your patience. I feel good about this one now.

**Comment by kjellr** [comment]
Thanks, folks! Visually, this is looking excellent, but I think we'll need to be a bit smarter about what happens when you click in the extra space. Right now, it selects the bottom-most block. Unfortunately, that ends up automatically jumping you back up to the very top of that block: 

scroll-select

This gets especially jarring on mobile. Would it be possible to add a brand new paragraph/placeholder block at the bottom if you click on the empty area?

**Comment by jasmussen** [comment]
That's a great observation. Hmmmmmmm. Could there be a regression involved? The empty area at the bottom is called the click redirector, and the _intent_ is to set focus in the text above. But if it sets focus in the top of the block, that seems off. @aduth can you recall the details? I think it was supposed to set focus at the bottom of the text above. Just like if you click below text in TextEdit.

**Comment by draganescu** [comment]
@kjellr @jasmussen I can't reproduce this in FF or Chrome. For me clicking on the redirector selects the last block but the cursor is at the end of the text.

**Comment by kjellr** [comment]
Ah, interesting. I'm able to reproduce in Chrome (74.0.3729.169) and Safari (12.1.1) for Mac. The cursor does appear at the end of the text, but I get scrolled up to the beginning of the block. 

chrome-mac

On Firefox, everything behaves as it should though: 

firefox-mac

**Comment by draganescu** [comment]
hi @kjellr I reproduced it easily when you mentioned mobile :) It is not a bug introduced by this PR, but a default behaviour of the browser that scrolls into view the focused element. Each browser does it differently though, Chrome is more eager and brings the top of the element into view, and in our case that leaves the caret under the fold.

In theory the fix I added is simple and straightforward, but in practice I am unsure if something else won't break. Maybe @ellatrix can take a look too: I added ` { preventScroll: true } ` to the focus call in `placeCaretAtHorizontalEdge`. I don't think this is quite right :/

`preventScroll` is not yet universal, but the big absentee, Firefox, handles focus correctly by default anyway. There is also a [polyfill](https://github.com/calvellido/focus-options-polyfill/blob/master/index.js) that implements a workaround, but I don't feel that is needed.

The main unintended effect is that if this method is used in a way where it should scroll, then this is the wrong way to do it, because that option would leave the focused element out of the view. .

File: packages/dom/src/dom.js [hunk 3]
```diff
@@ -264,7 +264,8 @@ export function placeCaretAtHorizontalEdge( container, isReverse ) {
 	}
 
 	if ( includes( [ 'INPUT', 'TEXTAREA' ], container.tagName ) ) {
-		container.focus();
+		container.focus( { preventScroll: true } );
```
Comment (talldan) [inline packages/dom/src/dom.js]: This would make it so that all other calls to `placeCaretAtHorizontalEdge` also prevent scroll. 

An option could be to make the function accept another parameter. 

`placeCaretAtHorizontalEdge( container, isReverse, preventScroll )` might be too specific and its heading towards this function potentiall having loads of params. Using an optional options object might be the best tradeoff `placeCaretAtHorizontalEdge( container, isReverse, { preventScroll: true } );`.

Seems like a legit option to me! Would be good to have someone who's more familiar with the DOM package also 👍 this idea.

**Review by talldan — DISMISSED** [review]
I found a separate issue. It seems like the editor no longer scrolls with the caret if I type past the end of the page:
itdoesntscrollasitype

My instinct is that it'll be solved by tidying up the `preventScroll` usage so that it's only applied by `WritingFlow`'s `focusLastTextField` method. Give it a go!

**Comment by draganescu** [comment]
Hey @talldan @jasmussen I think this works all right now, implemented Dan's suggestion and tested in both FF and Chrome. Thank you both for the help so far.

For now in both FF and Chrome being that we don't have the typewriter effect it feels as if the text is too close to the browser's bottom edge.

**Comment by mapk** [comment]
I just tested this myself. I focused on the recent changes and solution. Tested on both Chrome and FF on a Mac OSX.

1. Clicking below a large block focuses on the end of the block and doesn't cause Chrome or Firefox to jump upward.
2. I am not getting double scrollbars.
3. As one types, the cursor jumping to the middle of the screen feels a bit awkward as the page tries to keep the cursor in view. (This happens in Chrome but not in FF toward the end of the gif below)

clickable-area

File: packages/dom/src/dom.js [hunk 4]
```diff
@@ -257,14 +257,16 @@ export function computeCaretRect() {
  *
  * @param {Element} container Focusable element.
  * @param {boolean} isReverse True for end, false for start.
+ * @param {Object} focusOptions options that get passed to the focus() call.
  */
-export function placeCaretAtHorizontalEdge( container, isReverse ) {
+export function placeCaretAtHorizontalEdge( container, isReverse, focusOptions = {} ) {
```
Comment (talldan) [inline packages/dom/src/dom.js]: `focusOptions` is really specific, and I don't think that's the right way to go. It implies that other devs will have to add more params to this function if they want to add other options that are unrelated to focus.

`placeCaretAtHorizontalEdge( container, isReverse, options )` is more scaleable for future changes to the function. I also don't think this object should be passed straight into `container.focus`, `preventScroll` is the only thing that function cares about, so it can be destructured from a more general options object.

File: packages/dom/src/dom.js [hunk 5]
```diff
@@ -257,14 +257,16 @@ export function computeCaretRect() {
  *
  * @param {Element} container Focusable element.
  * @param {boolean} isReverse True for end, false for start.
+ * @param {Object} focusOptions options that get passed to the focus() call.
```
Comment (talldan) [inline packages/dom/src/dom.js]: The alignment of these params isn't right. Here's an example of the preferred alignment:
https://github.com/WordPress/gutenberg/blob/master/packages/rich-text/src/get-line-index.js

**Review by talldan — COMMENTED** [review]
Sorry for the delay in re-reviewing this. 

It's a shame, but I'm not completely convinced about the way this behaves at the moment. 

> As one types, the cursor jumping to the middle of the screen feels a bit awkward as the page tries to keep the cursor in view. (This happens in Chrome but not in FF toward the end of the gif below)

As mentioned by @mapk, this was the main issue for me. This change doesn't seem to prevent reaching the bottom of the page, instead you reach the bottom and the page jumps up. I think this might be a difficult experience for some users particularly those with impaired vision.

I see there's a lot of progress on https://github.com/WordPress/gutenberg/pull/16460 now, which introduces a typewriter like experience where the page scrolls as you type. That makes me wonder how this PR and that one work together. Maybe that PR solves the issue mentioned, but it might also make this PR no longer needed. Would be interesting to merge them together to see how it works!

**Comment by draganescu** [comment]
@mapk  @jasmussen in light of @talldan 's question,

> Maybe [#16460] solves the issue mentioned, but it might also make this PR no longer needed

I would say that the typewriter effect implemented by #16460 will be optional while this here would the default way the editor works. AM I right?

File: packages/block-editor/src/components/writing-flow/index.js [hunk 6]
```diff
@@ -341,7 +341,7 @@ class WritingFlow extends Component {
 		const focusableNodes = focus.focusable.find( this.container );
 		const target = findLast( focusableNodes, isTabbableTextField );
 		if ( target ) {
-			placeCaretAtHorizontalEdge( target, true );
+			placeCaretAtHorizontalEdge( target, true, { preventScroll: true } );
```
Comment (ellatrix) [inline packages/block-editor/src/components/writing-flow/index.js]: What does this fix?

Comment (draganescu) [reply packages/block-editor/src/components/writing-flow/index.js]: It is a workaround the default behaviour in Chrome which brings the focused element into view, which in our case scrolls up [the whole document](https://github.com/WordPress/gutenberg/pull/15165#issuecomment-498849250)

File: packages/dom/src/dom.js (hunk 4 above)
Comment (draganescu) [reply packages/dom/src/dom.js]: You are right :) of course.

File: packages/block-editor/src/components/writing-flow/index.js (hunk 6 above)
Comment (ellatrix) [reply packages/block-editor/src/components/writing-flow/index.js]: That seems... fine? Isn't the caret supposed to scroll into view? How is this related to the space at the bottom?

Comment (draganescu) [reply packages/block-editor/src/components/writing-flow/index.js]: Yes, except that it scrolls wrong in Chrome when clicking the bottom editor space, if we apply these changes: 

https://user-images.githubusercontent.com/1202812/58569372-fb0da900-8203-11e9-8be2-866ca3c9f0f5.gif

The caret is not into view because Chrome scrolls too much :)

**Comment by mapk** [comment]
> I would say that the typewriter effect implemented by #16460 will be optional while this here would the default way the editor works. Am I right?

It appears with @jasmussen's [suggestions on the typewriter experience](https://github.com/WordPress/gutenberg/pull/16460#issuecomment-521128687), it could be "on" by default. That being said, it sounds like there will still be the option to turn it off. In the case of it being "off", this PR should exist. The whole experience without either of these is difficult.

Does this sound right to everyone?

**Comment by jasmussen** [comment]
> Does this sound right to everyone?

Give or take. I would personally recommend launching a minimal version of the typewriter effect (as outlined in the comment you linked) for the plugin _without an opt-out_, especially if it's easy to flip a switch to turn it back off if something blows up. While I'm not a "decisions not options" absolutist, I think the setup with plugin-then-core allows us to test small iterations and potentially learning what needs to be tweaked or optional. Does that make sense?

**Comment by mapk** [comment]
> Does that make sense?

Yep! How 'bout we hold off on this PR until we get some feedback on the typewriter introduction?

**Comment by draganescu** [comment]
I think the typewriter feature made this pointless so I will close off this pr.
